### ISSUE TITLE
Better documentation?

### ISSUE THREAD

**Issue by jsumners** [pljson/pljson#53]
Do we need better documentation? We already have fairly thorough documentation, it just wasn't clear where to find it. I have added a link to it in the main readme to resolve that problem. But I'm thinking maybe we should create API documentation for each package and overhaul the examples.

Thoughts, @dsnz @boriborm?

**Comment by dsnz** [comment]
now I noticed new README.md in develop branch which should eventually be merged to to master branch to be useful to new users

doc/src/index.md needs surely rewrite so that does not refer to obscure version numbers and not to examples within doc, all examples should be from examples folder (which is updated and we avoid duplication and extra effort)
we could include a good short api documentation for each package in index.md
and perhaps for simplicity drop entirely the doc folder and move index.md to [root]/Documentation.md

**Comment by jsumners** [comment]
The documentation is where it is because it is used to generate http://pljson.github.io/pljson. Also, it keeps the root directory clean.

I'd rather be able to use something like [pldoc](http://pldoc.sourceforge.net/maven-site/) to generate the documentation. That would eliminate the need for `doc/src/`.

**Comment by jsumners** [comment]
By the way, I'm waiting on this bug to get fixed before I continue with my documentation update -- https://sourceforge.net/p/pldoc/bugs/109/

If they are unwilling to fix it then I will have to come up with a different plan.

**Comment by dsnz** [comment]
I do not track sourceforge so I got surprised by this. Do I understand well ? ... This is a sourceforge pldoc bug in how it parses source looking for documentation tags ? (I have no idea what is pldoc)
In this case we cannot do much since the keyword pipelined is necessary in the code to function.

However looking at the source code of pldoc, I suggest a hack (which I cannot test at all),
```
in file pldoc-0.8.3.1exp\src\javacc\PLSQL.jj
after line 683 [<PIPELINED>]
add  [<USING> <IDENTIFIER>]
```
(a similar change could be done in other versions)

**Comment by jsumners** [comment]
Yes, it's a bug in another project. I plan to use pldoc to generate html documentation from our documentation (to be written) . I will try out your suggested patch and possibly submit it to them.

**Comment by dsnz** [comment]
Hi, good :+1:

However I checked now the build.xml file and found that PLSQL.jj is generated by concatenating a few other files and if you build with supplied build.xml you will lose the proposed hack.
So you should make an extra change too

```
in file pldoc-0.8.3.1exp\src\javacc\Productions.jj
after line 286 	[<PIPELINED>]
add [<USING> <IDENTIFIER>]
```

Sorry for noticing this somewhat late but I am very hopeful that it will work.
This is a hack in the sense that it will not produce a java exception allowing you to continue, but it does not handle the exact correct Oracle syntax not the documentation actions necessary for this extra syntax.

**Comment by jsumners** [comment]
@dsnz can you write a working pared down test case of the pipelined function declaration? One of the pldoc guys needs a simple, working, example to resolve the issue. I was just looking at the documentation (http://docs.oracle.com/cd/B10501_01/appdev.920/a96595/dci12tbl.htm#1007111) and trying to write one, but I must admit that this is beyond my Oracle skills.

This is what I am being asked about the issue:

> Can you confirm that your type or type body compiles?

> The CREATE TYPE and CREATE TYPE BODY syntax diagrams have  little information about extended PIPELINED clauses 

> http://docs.oracle.com/database/121/LNPLS/create_type.htm#i2121991
> http://docs.oracle.com/database/121/LNPLS/create_type_body.htm#LNPLS2185

**Comment by dsnz** [comment]
sure ! here it is

```
create type num_array as table of number
/

create function num_generator(n in number default null) return num_array pipelined as
begin
  for i in 1 .. nvl(n, 999999999)
  loop
    pipe row(i);
  end loop;
  return;
end;
/

select *
from table(num_generator(3))
/
```

what about my fix proposal in PLSQL.jj and Productions.jj ? did you try it ?

**Comment by jsumners** [comment]
I attempted to try it, but the project is way too complicated to setup for compilation. Thank you for the simple test case. Hopefully this will help the pldoc authors to solve it.

**Comment by jsumners** [comment]
Resolved by #74
